src/Altinn.App.Core/Internal/Data/CachedFormDataAccessor.cs

```diff
@@ -0,0 +1,159 @@
+        Instance = instance;
+        _dataClient = dataClient;
+        _appMetadata = appMetadata;
+        _appModel = appModel;
+    }
+
+    public Instance Instance { get; }
+
+    public async Task<object?> GetSingleDataByType(string dataType)
+    {
+        var appMetadata = await _appMetadata.GetApplicationMetadata();
+        var dataTypeObj = appMetadata.DataTypes.Find(d => d.Id == dataType);
+        if (dataTypeObj == null)
+        {
+            throw new InvalidOperationException($"Data type {dataType} not found in app metadata");
+        }
+        if (dataTypeObj.MaxCount != 1)
+        {
+            throw new InvalidOperationException($"Data type {dataType} is not a single data type");
+        }
+        var dataElement = Instance.Data.Find(d => d.DataType == dataType);
+        if (dataElement == null)
+        {
+            return null;
+        }
+
+        return await GetData(dataElement);
+    }
+
+    /// <inheritdoc />
+    public async Task<object> GetData(DataElementId dataElementId)
+    {
+        return await _cache.GetOrCreate(
+            dataElementId,
+            async _ =>
+            {
+                var appMetadata = await _appMetadata.GetApplicationMetadata();
+                var dataElementIdString = dataElementId.Id.ToString();
+                var dataElementType = Instance.Data.Find(d => d.Id == dataElementIdString)?.DataType;
+                var dataType = appMetadata.DataTypes.Find(d => d.Id == dataElementType);
+                if (dataType == null)
+                {
+                    throw new InvalidOperationException(
+                        $"Data type {dataElementType ?? "unknown"} for data element id {dataElementId} not found in app metadata"
+                    );
+                }
+
+                if (dataType.AppLogic?.ClassRef != null)
+                {
+                    return await GetFormData(dataElementId, dataType);
+                }
+
+                return await GetBinaryData(dataElementId);
+            }
+        );
+    }
+
+    /// <summary>
+    /// Add data to the cache, so that it won't be fetched again
+    /// </summary>
+    public void Set(DataElementId dataElementId, object data)
+    {
+        _cache.Set(dataElementId, data);
+    }
+
+    /// <summary>
+    /// Simple wrapper around a ConcurrentDictionary using Lazy to ensure that the valueFactory is only called once
+    /// </summary>
+    /// <typeparam name="TKey">The type of the key in the cache</typeparam>
+    /// <typeparam name="TValue">The type of the object to cache</typeparam>
+    private sealed class LazyCache<TKey, TValue>
+        where TKey : notnull
+        where TValue : notnull
+    {
+        private readonly ConcurrentDictionary<TKey, Lazy<Task<TValue>>> _cache = new();
+
+        public async Task<TValue> GetOrCreate(TKey key, Func<TKey, Task<TValue>> valueFactory)
+        {
+            Task<TValue> task;
+            lock (_cache)
```

Comment:
What's the reasoning for using both a `ConcurrentDictionary` and a `lock`?

Reply:
https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-8.0#system-collections-concurrent-concurrentdictionary-2-addorupdate(-0-system-func((-0-1))-system-func((-0-1-1)))
> If you call [AddOrUpdate](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-8.0) simultaneously on different threads, addValueFactory may be called multiple times, but its key/value pair might not be added to the dictionary for every call.
For modifications and write operations to the dictionary, [ConcurrentDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-8.0) uses fine-grained locking to ensure thread safety (read operations on the dictionary are performed in a lock-free manner). The addValueFactory and updateValueFactory delegates may be executed multiple times to verify the value was added or updated as expected. However, they are called outside the locks to avoid the problems that can arise from executing unknown code under a lock. Therefore, [AddOrUpdate](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-8.0) is not atomic with regards to all other operations on the [ConcurrentDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-8.0) class.

Reply:
I suggested this in the maskinporten cache, but you wrote some wizard code that actually trigged that race condition. As I (think i) know the Lazy constructor is reasonably fast, there is no reason not to lock to be extra sure.

Reply:
Changed to using regular Dictionary, as it is protected by a lock